ldp/nn/handlers/transformer_handler.py

```diff
@@ -0,0 +1,1007 @@
+                cpu_offload=self.worker_config.offload_cpu,
+                activation_checkpointing=self.worker_config.activation_checkpointing,
+                cpu_ram_efficient_loading=self.worker_config.cpu_ram_efficient_loading,
+                sync_module_states=self.worker_config.cpu_ram_efficient_loading,
+                state_dict_type=self.worker_config.state_dict_type,
+                backward_prefetch=self.worker_config.backward_prefetch,
+            ),
+        )
+
+        if self.config.lm_config.device == "meta":
+            self.module = prepare_model_for_fsdp_with_meta_device(self.module)
+
+        self.module = self.accelerator.prepare(self.module)
+
+    def set_seed(self, seed: int) -> None:
+        """Set the seed for the current worker."""
+        set_seed(seed)
+
+    def _exec_func(
+        self,
+        func: Callable[Concatenate[Self, TParams], TReturn] | str,
+        *args,
+        **kwargs,
+    ) -> TReturn:
+        # data will be on CPU when sent from controller
+        data_device = _get_data_device()
+        to_device = partial(_move_tensor, device=data_device)
+        args = tree.map_structure(to_device, args)
+        kwargs = tree.map_structure(to_device, kwargs)
+
+        with torch.autocast(
+            device_type=self.module.device.type, dtype=self.module.dtype
+        ):
+            res = (
+                getattr(self, func)(*args, **kwargs)
+                if isinstance(func, str)
+                else func(self, *args, **kwargs)
+            )
+
+        # Needed to prevent GPU memory leak to the main process scheduling the workers
+        if isinstance(res, GenerateDecoderOnlyOutput):
+            res.past_key_values = None
+            res["past_key_values"] = None
+
+        to_cpu = partial(_move_tensor, device=torch.device("cpu"))
+        return tree.map_structure(to_cpu, res)
+
+    def __del__(self) -> None:
+        dist.destroy_process_group()
+
+
+class ParallelAsyncTransformer(AsyncTransformerInterface):
+    def __init__(self, config: TransformerHandlerConfig):
+        self._initialized = False
+
+        parallel_mode_config = config.parallel_mode_config
+        if not parallel_mode_config:
+            raise ValueError("Parallel mode config must be provided.")
+        self.config = config
+        self.tokenizer = config.lm_config.get_tokenizer()
+        maybe_set_tokenizer_chat_template(self.tokenizer)
+        self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        match parallel_mode_config.execution_mode:
+            case ExecutionMode.LOCAL_MACHINE:
+                self._init_local_cluster(config, parallel_mode_config)
+            case ExecutionMode.SLURM_CLUSTER:
+                self._init_slurm_cluster(config, parallel_mode_config)
+            case _:
+                raise NotImplementedError(
+                    f"Unsupported execution mode: {parallel_mode_config.execution_mode}"
+                )
```

Comment:
```suggestion
                typing.assert_never(parallel_mode_config.execution_mode)
```

Same as above, this `NotImplementedError` is misleading because actually, it's an impossible branch, not an unimplemented branch